In [3]:
!pip install h2o

In [17]:
import pandas as pd

In [4]:
import h2o

In [5]:
# Initialize local server
from h2o.estimators import H2ORandomForestEstimator

h2o.init()

Checking whether there is an H2O instance running at http://localhost:54321..... not found.
Attempting to start a local H2O server...
; Java HotSpot(TM) 64-Bit Server VM (build 25.201-b09, mixed mode)
  Starting server from C:\Users\navee\Anaconda3\lib\site-packages\h2o\backend\bin\h2o.jar
  Ice root: C:\Users\navee\AppData\Local\Temp\tmpjltjv2sl
  JVM stdout: C:\Users\navee\AppData\Local\Temp\tmpjltjv2sl\h2o_navee_started_from_python.out
  JVM stderr: C:\Users\navee\AppData\Local\Temp\tmpjltjv2sl\h2o_navee_started_from_python.err
  Server is running at http://127.0.0.1:54321
Connecting to H2O server at http://127.0.0.1:54321... successful.


--------------------------  ------------------------------------------
H2O cluster uptime:         02 secs
H2O cluster timezone:       America/New_York
H2O data parsing timezone:  UTC
H2O cluster version:        3.22.1.6
H2O cluster version age:    7 days, 22 hours and 37 minutes
H2O cluster name:           H2O_from_python_navee_qpu5y6
H2O cluster total nodes:    1
H2O cluster free memory:    3.500 Gb
H2O cluster total cores:    12
H2O cluster allowed cores:  12
H2O cluster status:         accepting new members, healthy
H2O connection url:         http://127.0.0.1:54321
H2O connection proxy:
H2O internal security:      False
H2O API Extensions:         Amazon S3, Algos, AutoML, Core V3, Core V4
Python version:             3.7.1 final
--------------------------  ------------------------------------------

In [7]:
# Load data from CSV
data = h2o.import_file('../../Data/finaldataset.csv')

Parse progress: |█████████████████████████████████████████████████████████| 100%


In [8]:
data

C1 dti revol_bal total_pymnt loan_status loan_amnt int_rate sub_grade annual_inc acc_now_delinq delinq_2yrs pub_rec open_acc inq_last_6mths revol_util emp_length addr_state_AL addr_state_AR addr_state_AZ addr_state_CA addr_state_CO addr_state_CT addr_state_DC addr_state_DE addr_state_FL addr_state_GA addr_state_HI addr_state_IA addr_state_ID addr_state_IL addr_state_IN addr_state_KS addr_state_KY addr_state_LA addr_state_MA addr_state_MD addr_state_ME addr_state_MI addr_state_MN addr_state_MO addr_state_MS addr_state_MT addr_state_NC addr_state_ND addr_state_NE addr_state_NH addr_state_NJ addr_state_NM addr_state_NV addr_state_NY addr_state_OH addr_state_OK addr_state_OR addr_state_PA addr_state_RI addr_state_SC addr_state_SD addr_state_TN addr_state_TX addr_state_UT addr_state_VA addr_state_VT addr_state_WA addr_state_WI addr_state_WV addr_state_WY application_type_JOINT purpose_credit_card purpose_debt_consolidation purpose_educational purpose_home_improvement purpose_house purpose_major_purchase purpose_medical purpose_moving purpose_other purpose_renewable_energy purpose_small_business purpose_vacation purpose_wedding term_ 60 months home_ownership_MORTGAGE home_ownership_NONE home_ownership_OTHER home_ownership_OWN home_ownership_RENT loan_income_ratio 
 
 
 0 27.65 13648 5861.07 Fully Paid 5000 10.65 7 24000 0 0 0 3 1 83.7 10 0 0 1 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 1 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 1 20.8333 
 18 18.44 0 7164.5 Fully Paid 6000 11.71 8 84000 0 2 0 4 0 37.73 1 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 1 0 0 0 0 0 0 0 0 0 0 0 0 0 1 0 0 0 0 0 0 0 1 0 0 0 0 7.14286 
 21 13.22 32135 14025.4 Charged Off 21000 12.42 9 105000 0 0 0 7 0 90.3 10 0 0 0 0 0 0 0 0 1 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 1 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 1 20 
 25 29.44 13707 15823.5 Fully Paid 15000 9.91 6 92000 0 0 0 8 0 93.9 2 0 0 0 0 0 0 0 0 0 0 0 0 0 1 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 1 0 0 0 0 0 0 0 0 0 0 0 0 0 1 0 0 0 0 16.3043 
 33 11.93 2224 5714.59 Fully Paid 5000 8.9 5 24044 0 0 0 8 0 29.3 2 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 1 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 1 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 1 20.7952 
 51 15.71 26088 29695.6 Fully Paid 25600 9.91 6 110000 0 0 0 11 0 62 9 0 0 0 1 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 1 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 1 23.2727 
 59 24 6021 9079.33 Fully Paid 7100 16.77 17 33000 0 0 0 4 2 91.2 1 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 1 0 0 0 0 0 0 0 0 0 1 0 0 0 0 0 0 0 0 0 0 0 0 1 0 0 0 0 21.5152 
 64 21.73 11170 8772.91 Charged Off 10000 10.65 7 45996 0 0 0 7 1 30.1 10 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 1 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 1 0 0 0 0 0 0 0 0 0 0 0 1 0 0 0 0 1 21.741 
 69 11.88 19397 18562.4 Fully Paid 15000 14.65 13 61000 0 0 0 8 1 79.5 2 0 0 0 0 0 0 0 0 1 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 1 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 1 0 24.5902 
 76 21.82 11382 13413.2 Fully Paid 11000 13.49 11 48000 0 0 0 15 3 65 10 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 1 0 0 0 0 0 0 0 0 0 1 0 0 0 0 0 0 0 0 0 0 0 0 1 0 0 0 0 22.9167

In [23]:
print(type(data))
df = pd.read_csv('../../Data/finaldataset.csv')
numerical_data = df.drop(columns=['Unnamed: 0','int_rate']).select_dtypes(include = ['int64','float'])
categorical_data = df.select_dtypes(include = ['object'])
categorical_features = categorical_data.columns.values
numerical_features = numerical_data.columns.values
categorical_features=categorical_features.tolist()

<class 'h2o.frame.H2OFrame'>


In [24]:
numerical_features

array(['dti', 'revol_bal', 'total_pymnt', 'loan_amnt', 'sub_grade',
       'annual_inc', 'acc_now_delinq', 'delinq_2yrs', 'pub_rec',
       'open_acc', 'inq_last_6mths', 'revol_util', 'emp_length',
       'addr_state_AL', 'addr_state_AR', 'addr_state_AZ', 'addr_state_CA',
       'addr_state_CO', 'addr_state_CT', 'addr_state_DC', 'addr_state_DE',
       'addr_state_FL', 'addr_state_GA', 'addr_state_HI', 'addr_state_IA',
       'addr_state_ID', 'addr_state_IL', 'addr_state_IN', 'addr_state_KS',
       'addr_state_KY', 'addr_state_LA', 'addr_state_MA', 'addr_state_MD',
       'addr_state_ME', 'addr_state_MI', 'addr_state_MN', 'addr_state_MO',
       'addr_state_MS', 'addr_state_MT', 'addr_state_NC', 'addr_state_ND',
       'addr_state_NE', 'addr_state_NH', 'addr_state_NJ', 'addr_state_NM',
       'addr_state_NV', 'addr_state_NY', 'addr_state_OH', 'addr_state_OK',
       'addr_state_OR', 'addr_state_PA', 'addr_state_RI', 'addr_state_SC',
       'addr_state_SD', 'addr_state_TN', 'addr_state

In [29]:
# Input parameters that are going to train
training_columns = numerical_features.tolist()
# Output parameter train against input parameters
response_column = 'int_rate'


In [30]:
# Split data into train and testing
train, test = data.split_frame(ratios=[0.7])


In [31]:
# Define model
'''
Model Documentation
http://h2o-release.s3.amazonaws.com/h2o/master/3420/docs-website/h2o-py/docs/modeling.html#h2orandomforestestimator
'''
model = H2ORandomForestEstimator(ntrees=50, max_depth=20, nfolds=10)

In [32]:
# Train model
model.train(x=training_columns, y=response_column, training_frame=train)

drf Model Build progress: |███████████████████████████████████████████████| 100%


In [33]:
# Model performance
performance = model.model_performance(test_data=test)

print(performance)


ModelMetricsRegression: drf
** Reported on test data. **

MSE: 0.30766515708515185
RMSE: 0.5546757224587641
MAE: 0.3593217019510062
RMSLE: 0.03867987400188445
Mean Residual Deviance: 0.30766515708515185



In [35]:
y_pred = model.predict(test)

drf prediction progress: |████████████████████████████████████████████████| 100%


In [38]:
model.mae

Model Details
H2ORandomForestEstimator :  Distributed Random Forest
Model Key:  DRF_model_python_1553209515802_1


ModelMetricsRegression: drf
** Reported on train data. **

MSE: 0.32456355148892363
RMSE: 0.5697047932823838
MAE: 0.36822221347402073
RMSLE: 0.04008752453204016
Mean Residual Deviance: 0.32456355148892363

ModelMetricsRegression: drf
** Reported on cross-validation data. **

MSE: 0.3126674486730859
RMSE: 0.5591667449635089
MAE: 0.3610129117241735
RMSLE: 0.03931389812978054
Mean Residual Deviance: 0.3126674486730859
Cross-Validation Metrics Summary: 


mean       sd           cv_1_valid    cv_2_valid    cv_3_valid    cv_4_valid    cv_5_valid    cv_6_valid    cv_7_valid    cv_8_valid    cv_9_valid    cv_10_valid
----------------------  ---------  -----------  ------------  ------------  ------------  ------------  ------------  ------------  ------------  ------------  ------------  -------------
mae                     0.361012   0.00289075   0.364202      0.361807      0.358903      0.36257       0.362089      0.364241      0.357721      0.367695      0.35864       0.352252
mean_residual_deviance  0.312672   0.00651627   0.320508      0.312695      0.315327      0.322471      0.310828      0.311914      0.301791      0.32834       0.307065      0.295776
mse                     0.312672   0.00651627   0.320508      0.312695      0.315327      0.322471      0.310828      0.311914      0.301791      0.32834       0.307065      0.295776
r2                      0.968552   0.000730976  0.967114      0.968049      0.968294      0.967535      0.968871      0.969266      0.969614      0.967199      0.969221      0.970354
residual_deviance       0.312672   0.00651627   0.320508      0.312695      0.315327      0.322471      0.310828      0.311914      0.301791      0.32834       0.307065      0.295776
rmse                    0.55911    0.00583396   0.566135      0.559191      0.56154       0.567865      0.55752       0.558492      0.549355      0.57301       0.554135      0.543853
rmsle                   0.0393058  0.000562075  0.0399953     0.0389345     0.0397397     0.0400758     0.0388874     0.0389953     0.0386585     0.0407019     0.0392755     0.0377946

Scoring History: 


timestamp            duration           number_of_trees    training_rmse       training_mae         training_deviance
---  -------------------  -----------------  -----------------  ------------------  -------------------  -------------------
     2019-03-21 19:38:09  17 min 28.480 sec  0.0                nan                 nan                  nan
     2019-03-21 19:38:12  17 min 31.498 sec  1.0                0.890384872400561   0.5190388614420088   0.7927852209997632
     2019-03-21 19:38:17  17 min 36.284 sec  3.0                0.7819779757699699  0.46147050667758016  0.6114895545892995
     2019-03-21 19:38:22  17 min 41.717 sec  5.0                0.7376096419616481  0.4463113442937576   0.5440679839147907
     2019-03-21 19:38:27  17 min 46.442 sec  7.0                0.688202099106484   0.42041690691739614  0.4736221292145708
---  ---                  ---                ---                ---                 ---                  ---
     2019-03-21 19:40:01  19 min 20.991 sec  43.0               0.5728824174526582  0.3702077559166166   0.3281942642264018
     2019-03-21 19:40:07  19 min 26.378 sec  45.0               0.5715799478626591  0.3690453802923796   0.32670363679868014
     2019-03-21 19:40:12  19 min 31.968 sec  47.0               0.5705787588411279  0.36841033216556995  0.32556012004068197
     2019-03-21 19:40:18  19 min 37.903 sec  49.0               0.5702840410522034  0.3688281562423565   0.3252238874788312
     2019-03-21 19:40:21  19 min 40.379 sec  50.0               0.5697047932823838  0.36822221347402073  0.32456355148892363


See the whole table with table.as_data_frame()
Variable Importances: 


variable                relative_importance    scaled_importance       percentage
----------------------  ---------------------  ----------------------  ----------------------
sub_grade               32523844.0             1.0                     0.837801983218875
total_pymnt             1251481.375            0.03847888875005058     0.032237689306850836
term_ 60 months         1243591.0              0.03823628596914928     0.032034436215877315
revol_util              1145533.75             0.03522135175657588     0.029508518353308883
inq_last_6mths          776890.0               0.023886782878432204    0.02001239406826917
---                     ---                    ---                     ---
addr_state_ND           108.45791625976562     3.334720098269e-06      2.79383511180961e-06
home_ownership_NONE     102.1131591796875      3.1396399263164432e-06  2.630396556861079e-06
application_type_JOINT  47.95595932006836      1.4744862052612341e-06  1.2353274669967352e-06
addr_state_ME           21.07860565185547      6.480969977551075e-07   5.429769500374279e-07
addr_state_ID           0.9516056776046753     2.925870870628562e-08   2.4513004180549457e-08


See the whole table with table.as_data_frame()


<bound method ModelBase.mae of >

### http://127.0.0.1:54321/flow/index.html# - Local H2O instance
#### H2O has given Distributed Random Forest as the best model with R2 -> ~96% and MAE on train data is ~0.368 where as ~0.361 on test data
#### No. of trees - 50
#### K-fold =10
#### for more info look in H2O local instance, we can do it by importing "DRF_model" file into H2O local instance

In [52]:
feature_importances =pd.read_excel('FeatureImportance_H2O.xlsx')

In [55]:
feature_importances[['variable','scaled_importance']]

variable  scaled_importance
0                    sub_grade             1.0000
1                  total_pymnt             0.0388
2              term_ 60 months             0.0380
3                   revol_util             0.0356
4               inq_last_6mths             0.0235
5                    loan_amnt             0.0136
6                   annual_inc             0.0072
7          purpose_credit_card             0.0056
8                          dti             0.0046
9                    revol_bal             0.0043
10           loan_income_ratio             0.0038
11                    open_acc             0.0029
12                     pub_rec             0.0022
13                  emp_length             0.0019
14                 delinq_2yrs             0.0012
15     home_ownership_MORTGAGE             0.0008
16               purpose_other             0.0007
17         home_ownership_RENT             0.0006
18  purpose_debt_consolidation             0.0005
19               addr_state_CA             0.0005
20               addr_state_NY             0.0004
21      purpose_small_business             0.0004
22               addr_state_FL             0.0004
23               addr_state_TX             0.0003
24               addr_state_NJ             0.0003
25    purpose_home_improvement             0.0003
26          home_ownership_OWN             0.0002
27               addr_state_VA             0.0002
28               addr_state_IL             0.0002
29               addr_state_GA             0.0002
..                         ...                ...
53               addr_state_UT             0.0001
54               addr_state_KY             0.0001
55               addr_state_KS             0.0001
56               addr_state_OK             0.0001
57               purpose_house             0.0001
58               addr_state_TN             0.0001
59              purpose_moving             0.0001
60               addr_state_AR             0.0001
61               addr_state_HI             0.0001
62               addr_state_DC             0.0001
63               addr_state_NM             0.0001
64               addr_state_NH             0.0001
65               addr_state_RI             0.0001
66              acc_now_delinq             0.0001
67               addr_state_WV             0.0000
68         purpose_educational             0.0000
69               addr_state_DE             0.0000
70               addr_state_MS             0.0000
71               addr_state_MT             0.0000
72        home_ownership_OTHER             0.0000
73    purpose_renewable_energy             0.0000
74               addr_state_WY             0.0000
75               addr_state_VT             0.0000
76               addr_state_SD             0.0000
77               addr_state_NE             0.0000
78         home_ownership_NONE             0.0000
79               addr_state_ND             0.0000
80               addr_state_ME             0.0000
81      application_type_JOINT             0.0000
82               addr_state_ID             0.0000

[83 rows x 2 columns]

<table width = '100%'>
<caption> Manual model vs AutoML approaches </caption>
 <tr>
    <th>Criteria</th>
   <th>TPOT</th>
   <th>H2o.ai</th>
   <th>AutoSKLearn</th>
    <th>Manual Model</th>
     
 </tr>
    <tr>
    <td>Best Model</td>
   <td>Decision Tree Regressor</td>
   <td>Distributed Random Forest</td>
   <td>Random Forest Regressor</td>
   <td>Random Forest Regressor</td>
  
 </tr>
 <tr>
    <td>Interpretability</td>
   <td>interpretable</td>
   <td>not interpretable</td>
   <td>not interpretable</td>
   <td>not interpretable</td>
 </tr>
 <tr>
    <td>Reproducibility</td>
   <td>reproducible</td>
   <td>non reproducible</td>
   <td>non reproducible</td>
   <td>non reproducible</td>
 </tr>
  <tr>
    <td>MAE Score</td>
   <td>-0.35332900035151543</td>
   <td>-0.36822221347402073</td>
   <td>- 2.275351307098047</td>
   <td>0.508</td>
 </tr>
 </table>

<br>
<br>

#### Interpretability <br>
  - Interpretability is the degree to which a human can consistently predict the model’s result. The higher the interpretability of a machine learning model, the easier it is for someone to comprehend why certain decisions or predictions have been made. A model is better interpretable than another model if its decisions are easier for a human to comprehend than decisions from the other model
  <br>
 <b> DECISION TREE </b>
 <br>
 The interpretation is simple: Starting from the root node, you go to the next nodes and the edges tell you which subsets you are looking at. Once you reach the leaf node, the node tells you the predicted outcome. All the edges are connected by ‘AND’<br>
 - Feature importance: The overall importance of a feature in a decision tree can be computed.<br>
 - Tree decomposition:  Individual predictions of a decision tree can be explained by decomposing the decision path into one component per feature. We can track a decision through the tree and explain a prediction by the contributions added at each decision node.
 <br>
 <b> RANDOM FOREST </b><br>
  Random forests are typically treated as a black box

#### Reproducibility
  - Reproducibility reduces or eliminates variations when rerunning failed jobs or prior experiments, making it essential in the context of fault tolerance and iterative refinement of models.<br>
 <b> RANDOM FOREST </b><br>
 Random Forest gets a different aggregate model each run because you take different bootstrap samples each run<br>
 <b> DECISION TREE </b><br>
 To make reproducible example you need to specify int number in random_state